packages/report-server/src/__tests__/reportBuilder/context/buildContext.test.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import pick from 'lodash.pick';
+
+import { buildContext, ReqContext } from '../../../reportBuilder/context/buildContext';
+
+describe('buildContext', () => {
+  const HIERARCHY = 'test_hierarchy';
+  const ENTITIES = {
+    test_hierarchy: [
+      { id: 'ouId1', code: 'AU', name: 'Australia' },
+      { id: 'ouId2', code: 'FJ', name: 'Fiji' },
+      { id: 'ouId3', code: 'KI', name: 'Kiribati' },
+      { id: 'ouId4', code: 'TO', name: 'Tonga' },
+    ],
+  };
+
+  const entityApiMock = {
+    getEntities: async (
+      hierarchyName: keyof typeof ENTITIES,
+      entityCodes: string[],
+      queryOptions: { fields?: string[] } = {},
+    ) => {
+      const entities = ENTITIES[hierarchyName]?.filter(e => entityCodes.includes(e.code));
+      const { fields } = queryOptions;
+      return fields ? entities.map(e => pick(e, fields)) : entities;
+    },
+  };
```

Comment:
Really nice to see the hand written mock over using a mocking framework. Give that JS lets us do it, I'm quite partial to hand written mocks as I find them much easier to reason about and extend/debug!

Reply:
I agree! Plus I am not a big fan of `jest`'s mocking api: too many options which is confusing, expects a `string` to the path of the mocked module (instead of importing it in the test)... I miss good ol' `sinon.createStubInstance`!